### PR TITLE
Allow experiments to be blocked from or granted to locales.

### LINKED ISSUE #1602: Allow experiments to be hidden from users with specific locales.
Basically, there's a collision between our upcoming German search experiment and Universal Search. We should decouple the timelines here by adding an optional flag to the YAML that lets us disable certain experiments for certain locales.

### PR DESCRIPTION
**Pull request by chuckharmston**
Closes #1602, #1670.

### CODE REVIEW

File: docs/experiments/experiment-content/template.yaml [hunk 1]
```diff
@@ -69,3 +69,7 @@ launch_date: '' #UTM time stamp, this will hide the experiment on production ser
 order: 0 # where your experiment will appear in the list...lower numbers appear higher
 incompatible: # if your experiment has incompatible add ons, repeat the following syntax, otherwise delete this field
   # 'add-on ID': 'Add-on Name'
+blocklist: [] # Prevent the experiment from being shown to users of these locales
```
Comment (pdehaan) [inline docs/experiments/experiment-content/template.yaml]: `locale_blocklist` and `locale_grantlist`?

Comment (chuckharmston) [reply docs/experiments/experiment-content/template.yaml]: Good idea. Updated.

File: frontend/src/app/reducers/browser.js [hunk 2]
```diff
@@ -5,9 +5,11 @@ import { isMinFirefoxVersion } from '../lib/utils';
 const userAgent = navigator.userAgent.toLowerCase();
 const isUserAgentFirefox = userAgent.indexOf('firefox') > -1;
 const isUserAgentMinFirefox = isMinFirefoxVersion(isUserAgentFirefox, userAgent, config.minFirefoxVersion);
+const locale = (navigator.language || '').split('-')[0];
```
Comment (pdehaan) [inline frontend/src/app/reducers/browser.js:8]: Should this be `navigator.language` or should we check the `navigator.languages` array?
Are we concerned if the user has the German language, but it isn't the primary? Should they still see the experiment?

Comment (chuckharmston) [reply frontend/src/app/reducers/browser.js:8]: In a meeting, @johngruen said we just wanted to check against the first-choice language.

`navigator.language` [is the first item in `navigator.languages`](https://github.com/mozilla/testpilot/pull/1688).

File: frontend/src/app/selectors/experiment.js [hunk 3]
```diff
@@ -25,6 +25,31 @@ export const launchedExperimentSelector = createSelector(
...
+
+// Return the user's primary language subtag.
+export const localeSelector = store => store.browser.locale;
+
+
+// Passed a locale and set of experiments, filters out the experiments that are
+// blockedlisted in that locale, or grantlisted and not available in that locale.
+export const l10nSelector = (locale, experiments) => experiments.filter(exp => {
+  if ('locale_blocklist' in exp) {
+    return exp.locale_blocklist.indexOf(locale) === -1;
```
Comment (pdehaan) [inline frontend/src/app/selectors/experiment.js]: Not sure if you want to use the fancy `Array#includes()` here instead. Doesn't really add anything, but you may be able to reduce to:

``` js
return !exp.locale_blocklist.includes(locale);
...
return exp.locale_grantlist.includes(locale);
```

File: frontend/src/app/reducers/browser.js (hunk 2 above)
Comment (lmorchard) [reply frontend/src/app/reducers/browser.js:8]: What are the `locale` values meant to be, here? For example, it looks like this collapses `en-US`, `en-GB`, & `en-NZ` into just `en` and `es-MX`, `es-ES`, & `es-CO` into just `es`. Is that what we want? They're definitely different locales, despite shared languages.

Actually, just noticed discussion in #1602, will copy this to there

Comment (chuckharmston) [reply frontend/src/app/reducers/browser.js:8]: @johngruen has said (trying to remember where—may have been in a meeting) that we only care about first-choice language, irrespective of region codes. That may change in the future, but would be an easy change to make.

Comment (lmorchard) [reply frontend/src/app/reducers/browser.js:8]: <strike>Could we do this with a glob or pattern or regex? Would be nice if we didn't have to change everything if/when we decide to mind locales in the future</strike>

Edit: Bah, actually, never mind. If we don't need it now, let's worry about it when we do. Not like we have 100s of experiments that will need updating.

Comment (chuckharmston) [reply frontend/src/app/reducers/browser.js:8]: Found it: https://github.com/mozilla/testpilot/issues/1602#issuecomment-253537100

**Review by lmorchard — APPROVED** [review]
Looks good, tests pass. Played with block/grant lists and changed languages in my browser, worked as expected

Would be good to have a test somewhere that asserts / documents the locale flattening logic, but I don't see a good place for it except maybe for the browser reducer with a mocked out global. Kind of a nit, though, and we've already made noise about it

**Pull request merged**